operator/internal/configwatcher/configwatcher.go

```diff
@@ -0,0 +1,242 @@
+
+	internalclient "github.com/redpanda-data/redpanda-operator/operator/pkg/client"
+)
+
+const (
+	defaultConfigPath     = "/var/lib/redpanda.yaml"
+	defaultUsersDirectory = "/etc/secret/users"
+)
+
+type Option func(c *ConfigWatcher)
+
+// ConfigWatcher replaces the old bash scripts we leveraged for waiting
+// for a cluster to become stable and then creating superusers
+type ConfigWatcher struct {
+	adminClient    *rpadmin.AdminAPI
+	configPath     string
+	usersDirectory string
+	watch          bool
+	fs             afero.Fs
+	log            logr.Logger
+
+	// for testing mostly
+	initialized chan struct{}
+}
+
+func WithRedpandaConfigPath(path string) Option {
+	return func(c *ConfigWatcher) {
+		c.configPath = path
+	}
+}
+
+func WithUsersDirectory(path string) Option {
+	return func(c *ConfigWatcher) {
+		c.usersDirectory = path
+	}
+}
+
+func WithFs(fs afero.Fs) Option {
+	return func(c *ConfigWatcher) {
+		c.fs = fs
+	}
+}
+
+func WithInitializedSignal(ch chan struct{}) Option {
+	return func(c *ConfigWatcher) {
+		c.initialized = ch
+	}
+}
+
+func NewConfigWatcher(log logr.Logger, watch bool, options ...Option) *ConfigWatcher {
+	watcher := &ConfigWatcher{
+		log:            log,
+		watch:          watch,
+		configPath:     defaultConfigPath,
+		usersDirectory: defaultUsersDirectory,
+		fs:             afero.NewOsFs(),
+		initialized:    make(chan struct{}),
+	}
+
+	for _, option := range options {
+		option(watcher)
+	}
+
+	return watcher
+}
+
+func (w *ConfigWatcher) Start(ctx context.Context) error {
+	params := rpkconfig.Params{ConfigFlag: w.configPath}
+
+	config, err := params.Load(w.fs)
+	if err != nil {
+		return fmt.Errorf("loading rpk config: %w", err)
+	}
+
+	factory := internalclient.NewFactory(&rest.Config{}, nil).WithFS(w.fs)
+	client, err := factory.RedpandaAdminClient(ctx, config.VirtualProfile())
+	if err != nil {
+		return fmt.Errorf("initializing Redpanda admin API client: %w", err)
+	}
+
+	w.adminClient = client
+
+	close(w.initialized)
+
+	w.syncInitial(ctx)
+
+	return w.watchFilesystem(ctx)
+}
+
+func (w *ConfigWatcher) syncInitial(ctx context.Context) {
+	files, err := os.ReadDir(w.usersDirectory)
+	if err != nil {
+		w.log.Error(err, "unable to get user directory files")
+		return
+	}
+
+	for _, file := range files {
+		if file.IsDir() {
+			continue
+		}
+		filePath := path.Join(w.usersDirectory, file.Name())
+		w.SyncUsers(ctx, filePath)
+	}
+}
+
+func (w *ConfigWatcher) watchFilesystem(ctx context.Context) error {
+	if !w.watch {
+		<-ctx.Done()
+		return nil
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+	defer watcher.Close()
+
+	if err := watcher.Add(w.usersDirectory); err != nil {
+		return err
+	}
+
+	for {
+		select {
+		case err := <-watcher.Errors:
+			// here we don't return as that'd crash the broker, instead
+			// just log the error and move on after some sleep time.
+			w.log.Error(err, "watcher returned an error")
+			time.Sleep(5 * time.Second)
```

Comment:
Should the watcher be re-instantiated?

Reply:
I don't believe that it needs to be re-initialized based on the `fsnotify` [usage docs](https://github.com/fsnotify/fsnotify). It's a bit hard to test with the in-memory filesystem, but do you think it's worth writing a test that just dumps some files to disk in a temporary directory and uses them instead of the in-memory `afero` FS?

Reply:
I don't think it's worth implementing test for that. I'm just worry that if watcher crash and we will not re-instantiate it, then we might miss SASL users update.